cpp/src/transform/codify.cpp

```diff
@@ -0,0 +1,56 @@
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/search.hpp>
+#include <cudf/detail/stream_compaction.hpp>
+#include <cudf/detail/transform.hpp>
+#include <cudf/table/table_view.hpp>
+
+namespace cudf {
+namespace detail {
+
+std::pair<std::unique_ptr<column>, std::unique_ptr<column>> codify(
+  column_view const& input_column, rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+{
+  // side effects of this function were are now dependent on:
+  // - resulting column elements are sorted ascending
+  // - nulls are sorted to the beginning
+  auto table_keys = cudf::detail::drop_duplicates(table_view{{input_column}},
+                                                  std::vector<size_type>{0},
+                                                  duplicate_keep_option::KEEP_FIRST,
+                                                  null_equality::EQUAL,
+                                                  mr,
+                                                  stream)
+                      ->release();  // true == nulls are equal
+  std::unique_ptr<column> keys_column(std::move(table_keys.front()));
+
+  if (input_column.has_nulls()) {
+    // the single null entry should be at the beginning -- side effect from drop_duplicates
+    // copy the column without the null entry
+    keys_column = std::make_unique<column>(
+      slice(keys_column->view(), std::vector<size_type>{1, keys_column->size()}).front(),
+      stream,
+      mr);
+    keys_column->set_null_mask(rmm::device_buffer{0, stream, mr}, 0);  // remove the null-mask
+  }
+
+  // this returns a column with no null entries
+  // - it appears to ignore the null entries in the input and tries to place the value regardless
+  auto indices_column = cudf::detail::lower_bound(table_view{{keys_column->view()}},
+                                                  table_view{{input_column}},
+                                                  std::vector<order>{order::ASCENDING},
+                                                  std::vector<null_order>{null_order::AFTER},
```

Comment:
We should make null placement configurable. One question to consider:

Let's say we want to codify {'a', 'b', null, 'b', 'a'}. There are three possible answers:

1. `{0, 1, 2, 1, 0}` (nulls come last)
2. `{1, 2, 0, 1, 2}` (nulls come first)
3. `{0, 1, null, 1, 0}` (nulls are not counted at all)

Is (3) interesting enough to support in libcudf? Or are we OK with telling the user to call `replace` as a post-processing step?

Reply:
> We should make null placement configurable

Should we?

Reply:
Could be a post-processing step. Marking resolved.